gtsfm/frontend/detector_descriptor/d2net.py

```diff
@@ -0,0 +1,117 @@
+"""Implements the D2Net detector-descriptor by wrapping around the original Pytorch source code.
+
+Paper: https://arxiv.org/abs/1905.03561
+https://github.com/mihaidusmanu/d2-net
+
+Authors: John Lambert
+"""
+
+from pathlib import Path
+from typing import Tuple
+
+import numpy as np
+import scipy
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+import thirdparty.d2net.lib.pyramid as d2net_pyramid
+from gtsfm.common.image import Image
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.frontend.detector_descriptor.detector_descriptor_base import DetectorDescriptorBase
+from thirdparty.d2net.lib.model_test import D2Net
+from thirdparty.d2net.lib.utils import preprocess_image
+
+MAX_EDGE_PX = 1600
+MAX_SUM_EDGES_PX = 2800
+USE_MULTISCALE = False
+USE_RELU = True
+PREPROCESSING_METHOD = "torch"
+
+MODEL_PATH = Path(__file__).resolve().parent.parent.parent.parent / "thirdparty" / "d2net" / "weights" / "d2_tf.pth"
+
+
+class D2NetDetDesc(DetectorDescriptorBase):
+    """D2-Net detector descriptor."""
+
+    def __init__(self, max_keypoints: int = 5000, model_path: Path = MODEL_PATH, use_cuda: bool = True) -> None:
+        """ """
+        super().__init__()
+        self.max_keypoints = max_keypoints
+        self.model_path = model_path
+        self.use_cuda = use_cuda and torch.cuda.is_available()
+        self.device = torch.device("cuda" if self.use_cuda else "cpu")
+
+    def detect_and_describe(self, image: Image) -> Tuple[Keypoints, np.ndarray]:
+        """Extract keypoints and their corresponding descriptors.
+
+        Adapted from:
+        https://github.com/mihaidusmanu/d2-net/blob/master/extract_features.py
+
+        Args:
+            image: the input image.
+
+        Returns:
+            Detected keypoints, with length N <= max_keypoints.
+            Corr. descriptors, of shape (N, D) where D is the dimension of each descriptor.
+        """
+        model = D2Net(model_file=self.model_path, use_relu=USE_RELU, use_cuda=self.use_cuda)
+        model.eval()
+
+        resized_image, fact_i, fact_j = resize_image(image.value_array)
```

Comment:
Yeah, it's the scale factor on y and on x. I updated the function docstrings.

Reply:
rescale APIs generally accept the scale factors, not define them. I think its better design to pass the max dimensions to it, rather than define it inside. But up to you.

Reply:
Agreed. This code closely mirrors the author's original implementation, though. I'll update the API to pass the max dimensions of it, I agree that part if hiding a side effect.